shared/slackutil/messaging.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// messaging.go includes functions to send message to Slack channel.
+
+package slackutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"net/http"
+	"net/url"
+)
+
+const (
+	slackChatPostMessageURL = "https://slack.com/api/chat.postMessage"
+)
+
+// SlackOperations defines the operations that can be done to Slack
+type SlackOperations interface {
+	PostMessageToChannel(text, channel string) error
+}
+
+// slackClient contains Slack bot related information
+type slackClient struct {
+	userName  string
+	tokenStr  string
+	iconEmoji *string
+}
+
+// NewSlackClient reads token file and stores it for later authentication
+func NewSlackClient(userName, slackTokenPath string) (SlackOperations, error) {
+	b, err := ioutil.ReadFile(slackTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	return &slackClient{
+		userName: userName,
+		tokenStr: strings.TrimSpace(string(b)),
+	}, nil
+}
+
+// PostMessageToChannel posts the given text to channel
+func (c *slackClient) PostMessageToChannel(text, channel string) error {
+	uv := &url.Values{}
+	uv.Add("username", c.userName)
+	uv.Add("token", c.tokenStr)
+	if nil != c.iconEmoji {
+		uv.Add("icon_emoji", *c.iconEmoji)
+	}
+	uv.Add("channel", channel)
+	uv.Add("text", text)
+
+	return c.postMessage(uv)
+}
+
+// postMessage does http post
+func (c *slackClient) postMessage(uv *url.Values) error {
+	resp, err := http.PostForm(slackChatPostMessageURL, *uv)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	t, _ := ioutil.ReadAll(resp.Body)
```

Comment:
do this after the if

Reply:
It's used in the following error message, I think it's better to keep it for debugging purpose.